File: packages/sdk/src/clients/contacts.ts
// MultiWA Gateway SDK - Contacts Client
// packages/sdk/src/clients/contacts.ts

import type { MultiWAClient } from '../client';
import type {
  Contact,
  CreateContactOptions,
  UpdateContactOptions,
  ImportContactsOptions,
  ImportCsvOptions,
  PaginatedResponse,
} from '../types';

export class ContactClient {
  constructor(private client: MultiWAClient) {}

  /**
   * Create a new contact
   */
  async create(options: CreateContactOptions): Promise<Contact> {
    return this.client.post('/contacts', options);
  }

  /**
   * List contacts with filtering
   */
  async list(profileId: string, options?: {
    search?: string;
    tags?: string[];
    limit?: number;
    offset?: number;
  }): Promise<PaginatedResponse<Contact>> {
    return this.client.get('/contacts', { profileId, ...options });
  }

  /**
   * Get a contact by ID
   */
  async get(id: string): Promise<Contact> {
    return this.client.get(`/contacts/${id}`);
  }

  /**
   * Update a contact
   */
  async update(id: string, data: UpdateContactOptions): Promise<Contact> {
    return this.client.put(`/contacts/${id}`, data);
  }

  /**
   * Delete a contact
   */
  async delete(id: string): Promise<{ success: boolean }> {
    return this.client.delete(`/contacts/${id}`);
  }

  /**
   * Bulk import contacts from JSON array
   */
  async import(options: ImportContactsOptions): Promise<{
    created: number;
    updated: number;
    failed: number;
    errors: string[];
  }> {
    return this.client.post('/contacts/import', options);
  }

  /**
   * Import contacts from CSV data
   */
  async importCsv(options: ImportCsvOptions): Promise<{
    created: number;
    updated: number;
    failed: number;
    errors: string[];
  }> {
    return this.client.post('/contacts/import/csv', options);
  }

  /**
   * Export contacts to CSV
   */
  async exportCsv(profileId: string, options?: { tags?: string[] }): Promise<{
    csv: string;
    count: number;
    filename: string;
  }> {
    return this.client.get('/contacts/export/csv', { profileId, ...options });
  }

  /**
   * Add tags to a contact
   */
  async addTags(id: string, tags: string[]): Promise<Contact> {
    return this.client.post(`/contacts/${id}/tags`, { tags });
  }

  /**
   * Remove tags from a contact
   */
  async removeTags(id: string, tags: string[]): Promise<Contact> {
    return this.client.delete(`/contacts/${id}/tags`, { tags });
  }

  /**
   * Validate a phone number
   */
  async validatePhone(profileId: string, phone: string): Promise<{
    phone: string;
    validFormat: boolean;
    isIndonesian: boolean;
    onWhatsApp: boolean | null;
  }> {
    return this.client.get(`/contacts/profile/${profileId}/validate/${phone}`);
  }

  /**
   * Bulk validate phone numbers
   */
  async validateBulk(profileId: string, phones: string[]): Promise<{
    total: number;
    valid: number;
    results: any[];
  }> {
    return this.client.post(`/contacts/profile/${profileId}/validate`, { phones });
  }
}

File: packages/sdk-python/multiwa/contacts.py
"""
MultiWA Python SDK - Contacts Client

Client for managing WhatsApp contacts.
"""

from typing import Optional, List, Dict, Any
import httpx

from .types import Contact


class ContactsClient:
    """Synchronous Contacts API client"""
    
    def __init__(self, http: httpx.Client):
        self._http = http
    
    def create(
        self,
        profile_id: str,
        phone: str,
        name: Optional[str] = None,
        email: Optional[str] = None,
        tags: Optional[List[str]] = None,
        metadata: Optional[Dict[str, Any]] = None,
    ) -> Contact:
        """Create a new contact"""
        response = self._http.post("/contacts", json={
            "profileId": profile_id,
            "phone": phone,
            "name": name,
            "email": email,
            "tags": tags or [],
            "metadata": metadata or {},
        })
        response.raise_for_status()
        return Contact(**response.json())
    
    def get(self, contact_id: str) -> Contact:
        """Get a contact by ID"""
        response = self._http.get(f"/contacts/{contact_id}")
        response.raise_for_status()
        return Contact(**response.json())
    
    def update(
        self,
        contact_id: str,
        name: Optional[str] = None,
        email: Optional[str] = None,
        tags: Optional[List[str]] = None,
        metadata: Optional[Dict[str, Any]] = None,
    ) -> Contact:
        """Update a contact"""
        data = {}
        if name is not None:
            data["name"] = name
        if email is not None:
            data["email"] = email
        if tags is not None:
            data["tags"] = tags
        if metadata is not None:
            data["metadata"] = metadata
            
        response = self._http.patch(f"/contacts/{contact_id}", json=data)
        response.raise_for_status()
        return Contact(**response.json())
    
    def delete(self, contact_id: str) -> Dict[str, bool]:
        """Delete a contact"""
        response = self._http.delete(f"/contacts/{contact_id}")
        response.raise_for_status()
        return response.json()
    
    def list(
        self,
        profile_id: str,
        limit: int = 50,
        offset: int = 0,
        search: Optional[str] = None,
        tags: Optional[List[str]] = None,
    ) -> List[Contact]:
        """List contacts for a profile"""
        params: Dict[str, Any] = {
            "limit": limit,
            "offset": offset,
        }
        if search:
            params["search"] = search
        if tags:
            params["tags"] = ",".join(tags)
            
        response = self._http.get(f"/contacts/profile/{profile_id}", params=params)
        response.raise_for_status()
        return [Contact(**c) for c in response.json()]
    
    def add_tags(
        self,
        contact_id: str,
        tags: List[str],
    ) -> Contact:
        """Add tags to a contact"""
        response = self._http.post(f"/contacts/{contact_id}/tags", json={
            "tags": tags,
        })
        response.raise_for_status()
        return Contact(**response.json())
    
    def remove_tags(
        self,
        contact_id: str,
        tags: List[str],
    ) -> Contact:
        """Remove tags from a contact"""
        response = self._http.delete(f"/contacts/{contact_id}/tags", json={
            "tags": tags,
        })
        response.raise_for_status()
        return Contact(**response.json())


class AsyncContactsClient:
    """Asynchronous Contacts API client"""
    
    def __init__(self, http: httpx.AsyncClient):
        self._http = http
    
    async def create(
        self,
        profile_id: str,
        phone: str,
        name: Optional[str] = None,
        email: Optional[str] = None,
        tags: Optional[List[str]] = None,
        metadata: Optional[Dict[str, Any]] = None,
    ) -> Contact:
        """Create a new contact"""
        response = await self._http.post("/contacts", json={
            "profileId": profile_id,
            "phone": phone,
            "name": name,
            "email": email,
            "tags": tags or [],
            "metadata": metadata or {},
        })
        response.raise_for_status()
        return Contact(**response.json())
    
    async def get(self, contact_id: str) -> Contact:
        """Get a contact by ID"""
        response = await self._http.get(f"/contacts/{contact_id}")
        response.raise_for_status()
        return Contact(**response.json())
    
    async def list(
        self,
        profile_id: str,
        limit: int = 50,
        offset: int = 0,
        search: Optional[str] = None,
    ) -> List[Contact]:
        """List contacts for a profile"""
        params: Dict[str, Any] = {
            "limit": limit,
            "offset": offset,
        }
        if search:
            params["search"] = search
            
        response = await self._http.get(f"/contacts/profile/{profile_id}", params=params)
        response.raise_for_status()
        return [Contact(**c) for c in response.json()]
